File: 2537-count-the-number-of-good-subarrays/2537-count-the-number-of-good-subarrays.py
class Solution:
    def countGood(self, nums: List[int], k: int) -> int:
        n, cnt = len(nums), Counter()
        ans, same, right = 0, 0, -1

        for left in range(n):
            while same < k and right + 1 < n:
                right += 1
                same += cnt[nums[right]]
                cnt[nums[right]] += 1
            
            if same >= k:
                ans += n - right
            cnt[nums[left]] -= 1
            same -= cnt[nums[left]]

        return ans

File: 2537-count-the-number-of-good-subarrays/2537-count-the-number-of-good-subarrays.cpp
typedef long long ll;

class Solution {
public:
    long long countGood(vector<int>& nums, int k) {
        int n = nums.size(), same = 0, right = -1;
        ll ans = 0;
        unordered_map<int, long long> counts;

        for (int left = 0; left < n; ++left){
            while (same < k && right + 1 < n){
                ++right;
                same += counts[nums[right]];
                ++counts[nums[right]];
            }

            if (same >= k)
                ans += n - right;

            --counts[nums[left]];
            same -= counts[nums[left]];
        }

        return ans;
    }
};